src/core/lib/uri/uri_parser.cc

```diff
@@ -71,11 +114,41 @@ bool IsPChar(char c) {
 // https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
 bool IsPathChar(char c) { return IsPChar(c) || c == '/'; }
 
-// Checks if this string is made up of pchars, '/', '?', and '%' exclusively.
+// Returns true for any character allowed in a URI query or fragment,
+// as defined in:
+// See https://tools.ietf.org/html/rfc3986#section-3.4
+bool IsQueryOrFragmentChar(char c) {
+  return IsPChar(c) || c == '/' || c == '?';
+}
+
+// Same as IsQueryOrFragmentChar(), but excludes '&' and '='.
+bool IsQueryKeyOrValueChar(char c) {
+  return c != '&' && c != '=' && IsQueryOrFragmentChar(c);
+}
+
+// Returns a copy of str, percent-encoding any character for which
+// is_allowed_char() returns false.
+std::string PercentEncode(absl::string_view str,
+                          std::function<bool(char)> is_allowed_char) {
+  std::string out;
+  for (char c : str) {
+    if (!is_allowed_char(c)) {
+      std::string hex = absl::BytesToHexString(absl::string_view(&c, 1));
```

Comment:
Upper case is preferred. See §6.2.2.1. The absl documentation does not define the casing, but the implementation returns lower-case.